internal/stats/aggregate.go

```diff
@@ -0,0 +1,212 @@
+package stats
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/edulinq/autograder/internal/log"
+	"github.com/edulinq/autograder/internal/util"
+)
+
+type AggregationQuery struct {
+	EnableAggregation bool     `json:"enable-aggregation,omitempty"`
+	GroupByFields     []string `json:"group-by,omitempty"`
+	AggregateField    string   `json:"aggregate,omitempty"`
+}
+
+type QueryResponse struct {
+	Response []map[string]any `json:"response"`
+}
+
+const (
+	COUNT_KEY  = "count"
+	MAX_KEY    = "max"
+	MEAN_KEY   = "mean"
+	MEDIAN_KEY = "median"
+	MIN_KEY    = "min"
+)
+
+// Aggregate metrics based on groupByFields and the aggregateField.
+func ApplyAggregation(metrics []map[string]any, metricType any, groupByFields []string, aggregateField string) ([]map[string]any, error) {
+	err := validate(metricType, groupByFields, aggregateField)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to validate data for aggregation: '%v'.", err)
+	}
+
+	// Group together metrics based on their groupByFields.
+	groupedMetricBuckets, err := groupTogetherMetrics(metrics, groupByFields, aggregateField)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to group metrics: '%v'.", err)
+	}
+
+	aggregatedResults := make([]map[string]any, 0, len(groupedMetricBuckets))
+
+	// Aggregate on metrics that are grouped together.
+	for _, groupedMetrics := range groupedMetricBuckets {
+		result, err := computeGroupAggregation(groupedMetrics, aggregateField, groupByFields)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to compute aggregation for metrics '%v': '%v'.", groupedMetrics, err)
+		}
+
+		if result != nil {
+			aggregatedResults = append(aggregatedResults, result)
+		}
+	}
+
+	return aggregatedResults, nil
+}
+
+func extractJSONTags(reflectType reflect.Type, jsonTags map[string]bool) {
```

Comment:
I would take a minute and think about if you really want to do this reflexively instead of via a JSON conversion.
You already did a JSON conversion to generalize the input to aggregation,
so then mixing reflection seems a bit janky.